transcoder/src/extract.go

```diff
@@ -15,48 +21,86 @@ func (s *MetadataService) ExtractSubs(info *MediaInfo) (interface{}, error) {
 		return get_running()
 	}
 
-	err := extractSubs(info)
+	err := s.extractSubs(info)
 	if err != nil {
 		return set(nil, err)
 	}
 	_, err = s.database.Exec(`update info set ver_extract = $2 where sha = $1`, info.Sha, ExtractVersion)
 	return set(nil, err)
 }
 
-func (s *MetadataService) GetAttachmentPath(sha string, is_sub bool, name string) (string, error) {
+func (s *MetadataService) GetAttachment(sha string, name string) (io.ReadCloser, error) {
 	_, err := s.extractLock.WaitFor(sha)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	dir := "att"
-	if is_sub {
-		dir = "sub"
+
+	itemPath := fmt.Sprintf("%s/%s/%s", sha, "att", name)
+
+	item, err := s.storage.GetItem(context.TODO(), itemPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get attachment with path %q: %w", itemPath, err)
 	}
-	return fmt.Sprintf("%s/%s/%s/%s", Settings.Metadata, sha, dir, name), nil
+	return item, nil
 }
 
-func extractSubs(info *MediaInfo) error {
-	defer printExecTime("extraction of %s", info.Path)()
+func (s *MetadataService) GetSubtitle(sha string, name string) (io.ReadCloser, error) {
+	_, err := s.extractLock.WaitFor(sha)
+	if err != nil {
+		return nil, err
+	}
+
+	itemPath := fmt.Sprintf("%s/%s/%s", sha, "sub", name)
 
-	attachment_path := fmt.Sprintf("%s/%s/att", Settings.Metadata, info.Sha)
-	subs_path := fmt.Sprintf("%s/%s/sub", Settings.Metadata, info.Sha)
+	item, err := s.storage.GetItem(context.TODO(), itemPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get subtitle with path %q: %w", itemPath, err)
+	}
+	return item, nil
+}
+
+func (s *MetadataService) extractSubs(info *MediaInfo) (err error) {
+	defer printExecTime("extraction of %s", info.Path)()
 
-	os.MkdirAll(attachment_path, 0o755)
-	os.MkdirAll(subs_path, 0o755)
+	ctx := context.TODO()
 
 	// If there is no subtitles, there is nothing to extract (also fonts would be useless).
 	if len(info.Subtitles) == 0 {
 		return nil
 	}
 
+	// Create a temporary directory for writing attachments
+	// TODO if the transcoder ever uses the ffmpeg library directly, remove this
+	// and write directly to storage via stream instead
+	tempWorkingDirectory := filepath.Join(os.TempDir(), info.Sha)
+	if err := os.MkdirAll(tempWorkingDirectory, 0660); err != nil {
+		return fmt.Errorf("failed to create temporary directory: %w", err)
+	}
+
+	attachmentWorkingDirectory := filepath.Join(tempWorkingDirectory, "att")
+	if err := os.MkdirAll(attachmentWorkingDirectory, 0660); err != nil {
+		return fmt.Errorf("failed to create attachment directory: %w", err)
+	}
+
+	subtitlesWorkingDirectory := filepath.Join(tempWorkingDirectory, "sub")
+	if err := os.MkdirAll(subtitlesWorkingDirectory, 0660); err != nil {
+		return fmt.Errorf("failed to create subtitles directory: %w", err)
+	}
+
+	defer func() {
+		cleanupErr := os.RemoveAll(tempWorkingDirectory)
+		err = errors.Join(err, cleanupErr)
```

Comment:
`err` is assigned but can never be read. I think it's fine to just `os.RemoveAll()` without checking for the error.

Reply:
This is actually read. `err` here refers to the named return value, so both the error value from the "normal" logic and the cleanup error will be returned.